File: Python3/962. Maximum Width Ramp.py
class Solution:
    def maxWidthRamp(self, nums: List[int]) -> int:
        N = len(nums)
        stack = deque()

        for i, n in enumerate(nums):
            if not stack or nums[stack[-1]] > n:
                stack.append(i)

        ans = 0
        for r in range(N - 1, -1, -1):
            while stack and nums[stack[-1]] <= nums[r]:
                ans = max(ans, r - stack.pop())
            if ans >= r:
                break

        return ans

File: Java/962. Maximum Width Ramp.java
class Solution {
    public int maxWidthRamp(int[] nums) {
        Stack<Integer> stack = new Stack<>();
        int N = nums.length, ans = 0;
        for (int i = 0; i < N; i++) {
            if (stack.isEmpty() || nums[stack.peek()] > nums[i]) {
                stack.push(i);
            }
        }

        for (int r = N - 1; r >= 0; r--) {
            while (!stack.isEmpty() && nums[stack.peek()] <= nums[r]) {
                ans = Math.max(ans, r - stack.pop());
            }
            if (r <= ans) {
                break;
            }
        }

        return ans;
    }
}